packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesDevLauncherController.java

```diff
@@ -0,0 +1,116 @@
+    setDevelopmentSelectionPolicy();
+  }
+
+  private void setDevelopmentSelectionPolicy() {
+    UpdatesController controller = UpdatesController.getInstance();
+    SelectionPolicy currentSelectionPolicy = controller.getSelectionPolicy();
+    controller.setDefaultSelectionPolicy(new SelectionPolicy(
+            currentSelectionPolicy.getLauncherSelectionPolicy(),
+            currentSelectionPolicy.getLoaderSelectionPolicy(),
+            new ReaperSelectionPolicyDevelopmentClient()
+    ));
+  }
+
+  private void reset() {
+    mRawManifest = null;
+  }
+
+  @Override
+  public void fetchUpdateWithConfiguration(HashMap<String, Object> configuration, Context context, UpdateCallback callback) {
+    reset();
+
+    UpdatesController controller = UpdatesController.getInstance();
+    UpdatesConfiguration updatesConfiguration = new UpdatesConfiguration()
+            .loadValuesFromMetadata(context)
+            .loadValuesFromMap(configuration);
+    if (updatesConfiguration.getUpdateUrl() == null) {
+      callback.onFailure(new Exception("Failed to load update: UpdatesConfiguration object must include a valid update URL"));
+      return;
+    }
+    if (controller.getUpdatesDirectory() == null) {
+      callback.onFailure(controller.getUpdatesDirectoryException());
+      return;
+    }
+
+    DatabaseHolder databaseHolder = controller.getDatabaseHolder();
+    RemoteLoader loader = new RemoteLoader(context, updatesConfiguration, databaseHolder.getDatabase(), controller.getFileDownloader(), controller.getUpdatesDirectory());
+    loader.start(new RemoteLoader.LoaderCallback() {
+      @Override
+      public void onFailure(Exception e) {
+        databaseHolder.releaseDatabase();
+        callback.onFailure(e);
+      }
+
+      @Override
+      public void onSuccess(@Nullable UpdateEntity update) {
+        databaseHolder.releaseDatabase();
+        if (update != null) {
+          DatabaseLauncher launcher = new DatabaseLauncher(updatesConfiguration, controller.getUpdatesDirectory(), controller.getFileDownloader(), controller.getSelectionPolicy());
+          launcher.launch(databaseHolder.getDatabase(), context, new Launcher.LauncherCallback() {
+            @Override
+            public void onFailure(Exception e) {
+              databaseHolder.releaseDatabase();
+              callback.onFailure(e);
+            }
+
+            @Override
+            public void onSuccess() {
+              databaseHolder.releaseDatabase();
+              controller.setLauncher(launcher);
+              controller.setUpdatesConfiguration(updatesConfiguration);
+              callback.onSuccess(new Update() {
+                @Override
+                public JSONObject getManifest() {
+                  return mRawManifest;
+                }
+
+                @Override
+                public String getLaunchAssetPath() {
+                  return launcher.getLaunchAssetFile();
+                }
+              });
+            }
```

Comment:
Is there a closure issue requiring this code block to be inline?

Factoring this out to another function called `launchNewestUpdate` would reduce the double callback nesting. Additionally the function name would double as a descriptive comment.